File: frontend/src/services/reportService.js
import api from "./api";

/**
 * reportService.js
 * ----------------
 * Handles all calls to /api/v1/reports/*
 * All endpoints require ADMIN role.
 *
 * Endpoints covered (from ReportController.java):
 *   GET /reports/teaching-load              → getTeachingLoad()
 *   GET /reports/teaching-load/:teacherId   → getTeachingLoadByTeacher()
 *   GET /reports/room-utilisation           → getRoomUtilisation()
 *   GET /reports/conflicts                  → getConflictSummary()
 *
 * Note: These endpoints return Object[][] (raw JPQL projection rows).
 * Column order per report is documented on each function below.
 */

const unwrap = (res) => res.data?.data ?? res.data;

// ── Teaching Load ─────────────────────────────────────────────────────────────

/**
 * Get teaching load summary for ALL teachers in a term.
 *
 * Returns Object[][] — each row:
 * [teacherName, schoolId, departmentCode, classCount, unitTotal]
 *
 * @param {string} semester   - "FIRST" | "SECOND" | "SUMMER"
 * @param {string} schoolYear - e.g. "2024-2025"
 * @returns {Promise<Object[][]>}
 *
 * @example
 * const rows = await reportService.getTeachingLoad("FIRST", "2024-2025");
 * rows.forEach(([name, sid, dept, classes, units]) => { ... });
 */
const getTeachingLoad = async (semester, schoolYear) => {
  try {
    const res = await api.get("/reports/teaching-load", {
      params: { semester, schoolYear },
    });
    return unwrap(res) ?? [];
  } catch (err) {
    throw new Error(err?.response?.data?.message ?? "Failed to load teaching load report.");
  }
};

/**
 * Get teaching load for a single teacher, broken down by year level.
 *
 * Returns Object[][] — each row:
 * [yearLevel, classCount, unitTotal]
 *
 * @param {number} teacherId
 * @param {string} semester
 * @param {string} schoolYear
 * @returns {Promise<Object[][]>}
 */
const getTeachingLoadByTeacher = async (teacherId, semester, schoolYear) => {
  try {
    const res = await api.get(`/reports/teaching-load/${teacherId}`, {
      params: { semester, schoolYear },
    });
    return unwrap(res) ?? [];
  } catch (err) {
    throw new Error(err?.response?.data?.message ?? "Failed to load teacher load report.");
  }
};

// ── Room Utilisation ──────────────────────────────────────────────────────────

/**
 * Get room utilisation report for a specific campus and term.
 *
 * Returns Object[][] — each row:
 * [roomName, roomNumber, roomType, capacity, classCount]
 *
 * @param {number} campusId
 * @param {string} semester
 * @param {string} schoolYear
 * @returns {Promise<Object[][]>}
 *
 * @example
 * const rows = await reportService.getRoomUtilisation(1, "FIRST", "2024-2025");
 * rows.forEach(([name, num, type, cap, count]) => { ... });
 */
const getRoomUtilisation = async (campusId, semester, schoolYear) => {
  try {
    const res = await api.get("/reports/room-utilisation", {
      params: { campusId, semester, schoolYear },
    });
    return unwrap(res) ?? [];
  } catch (err) {
    throw new Error(err?.response?.data?.message ?? "Failed to load room utilisation report.");
  }
};

// ── Conflict Summary ──────────────────────────────────────────────────────────

/**
 * Get conflict summary for a term.
 *
 * Returns:
 * {
 *   totalUnresolved,   // across ALL terms
 *   termUnresolved,    // for this specific term
 *   semester,
 *   schoolYear
 * }
 *
 * @param {string} semester
 * @param {string} schoolYear
 * @returns {Promise<{ totalUnresolved: number, termUnresolved: number, semester: string, schoolYear: string }>}
 */
const getConflictSummary = async (semester, schoolYear) => {
  try {
    const res = await api.get("/reports/conflicts", {
      params: { semester, schoolYear },
    });
    return unwrap(res);
  } catch (err) {
    throw new Error(err?.response?.data?.message ?? "Failed to load conflict summary.");
  }
};

// ── Export ────────────────────────────────────────────────────────────────────

const reportService = {
  getTeachingLoad,
  getTeachingLoadByTeacher,
  getRoomUtilisation,
  getConflictSummary,
};

export default reportService;

File: src/main/java/com/timecraft/timecraft/service/ReportService.java
package com.timecraft.timecraft.service;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.timecraft.timecraft.model.CourseSubject.Semester;
import com.timecraft.timecraft.repository.ConflictLogRepository;
import com.timecraft.timecraft.repository.RoomRepository;
import com.timecraft.timecraft.repository.ScheduleRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ReportService {

    private final ScheduleRepository scheduleRepository;
    private final ConflictLogRepository conflictLogRepository;
    private final RoomRepository roomRepository;

    /**
     * Teaching load report — how many classes each teacher has in the term.
     * Returns: [teacher_id, teacher_full_name, class_count]
     */
    public List<Object[]> getTeachingLoadReport(Semester semester,
            String schoolYear) {
        return scheduleRepository.getTeachingLoadReport(semester, schoolYear);
    }

    /**
     * Teaching load by year level for a single teacher.
     * Returns: [year_level, class_count]
     */
    public List<Object[]> getTeacherLoadByYearLevel(Long teacherId,
            Semester semester,
            String schoolYear) {
        return scheduleRepository.countClassesByYearLevelForTeacher(
                teacherId, semester, schoolYear);
    }

    /**
     * Room utilisation report per campus.
     * Returns: [Room, scheduled_count]
     */
    public List<Object[]> getRoomUtilisationReport(Long campusId,
            Semester semester,
            String schoolYear) {
        return roomRepository.getRoomUtilisationByCampusAndTerm(
                campusId, semester, schoolYear);
    }

    /**
     * Total unresolved conflicts across the system.
     */
    public long getTotalUnresolvedConflicts() {
        return conflictLogRepository.countByResolvedFalse();
    }

    /**
     * Conflict summary for a specific term.
     */
    public long getUnresolvedConflictsByTerm(Semester semester,
            String schoolYear) {
        return conflictLogRepository
                .findUnresolvedByTerm(semester, schoolYear).size();
    }
}